djstripe/stripe_objects.py

```diff
@@ -0,0 +1,529 @@
+dj-stripe functionality.
+"""
+from contextlib import contextmanager
+import datetime
+import decimal
+
+from django.conf import settings
+from django.db import models
+from django.utils import timezone
+from django.utils.encoding import python_2_unicode_compatible, smart_text
+from jsonfield import JSONField
+
+from model_utils.models import TimeStampedModel
+import stripe
+from djstripe.managers import TransferManager, StripeObjectManager
+
+
+stripe.api_key = settings.STRIPE_SECRET_KEY
+stripe.api_version = getattr(settings, "STRIPE_API_VERSION", "2012-11-07")
+
+
+@contextmanager
+def stripe_temporary_api_key(temp_key):
+    """
+    A contextmanager
+
+    Temporarily replace the global api_key used in stripe API calls with the given value
+    the original value is restored as soon as context exits
+    """
+    import stripe
+    bkp_key = stripe.api_key
+    stripe.api_key = temp_key
+    yield
+    stripe.api_key = bkp_key
+
+
+def convert_tstamp(response, field_name=None):
+    # Overrides the set timezone to UTC - I think...
+    tz = timezone.utc if settings.USE_TZ else None
+
+    if not field_name:
+        return datetime.datetime.fromtimestamp(response, tz)
+    else:
+        if field_name in response and response[field_name]:
+            return datetime.datetime.fromtimestamp(response[field_name], tz)
+
+
+@python_2_unicode_compatible
+class StripeObject(TimeStampedModel):
+    # This must be defined in descendants of this model/mixin
+    # e.g. "Event", "Charge", "Customer", etc.
+    stripe_api_name = None
+    stripe_objects = StripeObjectManager()
+
+    stripe_id = models.CharField(max_length=50, unique=True)
+    # livemode = models.BooleanField(default=False)
+
+    class Meta:
+        abstract = True
+
+    @classmethod
+    def api(cls):
+        """
+        Get the api object for this type of stripe object (requires
+        stripe_api_name attribute to be set on model).
+        """
+        if cls.stripe_api_name is None:
+            raise NotImplementedError("StripeObject descendants are required to define "
+                                      "the stripe_api_name attribute")
+        # e.g. stripe.Event, stripe.Charge, etc
+        return getattr(stripe, cls.stripe_api_name)
+
+    def api_retrieve(self):
+        """
+        Implement very commonly used API function 'retrieve'
+        """
+        # Run stripe.X.retreive(id)
+        return type(self).api().retrieve(self.stripe_id)
+
+    @classmethod
+    def api_create(cls, **kwargs):
+        """
+        Call the stripe API's create operation for this model
+        """
+        return cls.api().create(**kwargs)
+
+    def str_parts(self):
+        """
+        extend this to add information to the string representation of the object
+        :rtype: list of str
+        """
+        return ["stripe_id={id}".format(id=self.stripe_id)]
+
+    @classmethod
+    def stripe_obj_to_record(cls, data):
```

Comment:
Also, please add some sort of docstring that mentions what this should be used for.